File: python/src/trig_reciprocal.py
"""
Trigonometric reciprocal identity functions that return radians.

Version: 1.2.0
"""

import math


def csc(angle_in_radians: float) -> float:
	"""The reciprocal of sin.

	Parameters
	----------
	angle_in_radians : float
		An angle in radians.

	Returns
	-------
	float
		The cosecant of angle_in_radians in radians.

	Raises
	------
	ZeroDivisionError
		If angle_in_radians is 0.
	"""
	return 1.0 / math.sin(angle_in_radians)


def sec(angle_in_radians: float) -> float:
	"""The reciprocal of cos.

	Parameters
	----------
	angle_in_radians : float
		An angle in radians.

	Returns
	-------
	float
		The secant of angle_in_radians in radians.
	"""
	return 1.0 / math.cos(angle_in_radians)


def cot(angle_in_radians: float) -> float:
	"""The reciprocal of tan.

	Parameters
	----------
	angle_in_radians : float
		An angle in radians.

	Returns
	-------
	float
		The cotangent of angle_in_radians in radians.

	Raises
	------
	ZeroDivisionError
		If angle_in_radians is 0.
	"""
	return 1.0 / math.tan(angle_in_radians)


def acsc(angle_in_radians: float) -> float:
	"""The inverse of csc.

	Parameters
	----------
	angle_in_radians: float
		An angle in radians.

	Returns
	-------
	float
		The arccosecant of angle_in_radians in radians.

	Raises
	------
	ZeroDivisionError
		If angle_in_radians is 0.

	ValueError
		If angle_in_radians is between 0 and 1.
	"""
	return math.asin(1.0 / angle_in_radians)


def asec(angle_in_radians: float) -> float:
	"""The inverse of sec.

	Parameters
	----------
	angle_in_radians: float
		An angle in radians.

	Returns
	-------
	float
		The arcsecant of angle_in_radians in radians.

	Raises
	------
	ZeroDivisionError
		If angle_in_radians is 0.

	ValueError
		If angle_in_radians is between 0 and 1.
	"""
	return math.acos(1.0 / angle_in_radians)


def acot(angle_in_radians: float) -> float:
	"""The inverse of cot.

	Parameters
	----------
	angle_in_radians: float
		An angle in radians.

	Returns
	-------
	float
		The arccotangent of angle_in_radians in radians.
	"""
	return math.atan(1.0 / angle_in_radians) if angle_in_radians != 0 else 1.5707963267948966


def csch(angle_in_radians: float) -> float:
	"""The hyperbolic version of csc.

	Parameters
	----------
	angle_in_radians : float
		An angle in radians.

	Returns
	-------
	float
		The hyperbolic cosecant of angle_in_radians in radians.

	Raises
	------
	ZeroDivisionError
		If angle_in_radians is 0.
	"""
	return 1.0 / math.sinh(angle_in_radians)


def sech(angle_in_radians: float) -> float:
	"""The hyperbolic version of sec.

	Parameters
	----------
	angle_in_radians : float
		An angle in radians.

	Returns
	-------
	float
		The hyperbolic secant of angle_in_radians in radians.

	Raises
	------
	ZeroDivisionError
		If angle_in_radians is 0.
	"""
	return 1.0 / math.cosh(angle_in_radians)


def coth(angle_in_radians: float) -> float:
	"""The hyperbolic version of cot.

	Parameters
	----------
	angle_in_radians : float
		An angle in radians.

	Returns
	-------
	float
		The hyperbolic cotangent of angle_in_radians in radians.

	Raises
	------
	ZeroDivisionError
		If angle_in_radians is 0.
	"""
	return 1.0 / math.tanh(angle_in_radians)


def acsch(angle_in_radians: float) -> float:
	"""The inverse of csch.

	Parameters
	----------
	angle_in_radians: float
		An angle in radians.

	Returns
	-------
	float
		The hyperbolic arccosecant of angle_in_radians in radians.

	Raises
	------
	ZeroDivisionError
		If angle_in_radians is 0.
	"""
	return math.log((1.0 / angle_in_radians) + (math.sqrt(1.0 + math.pow(angle_in_radians, 2.0)) / abs(angle_in_radians)))


def asech(angle_in_radians: float) -> float:
	"""The inverse of sech.

	Parameters
	----------
	angle_in_radians: float
		An angle in radians.

	Returns
	-------
	float
		The hyperbolic arcsecant of angle_in_radians in radians.

	Raises
	------
	ZeroDivisionError
		If angle_in_radians is 0.

	ValueError
		If angle_in_radians is not between 0 and 1.
	"""
	return math.log((1.0 + math.sqrt(1.0 - math.pow(angle_in_radians, 2.0))) / angle_in_radians)


def acoth(angle_in_radians: float) -> float:
	"""The inverse of coth.

	Parameters
	----------
	angle_in_radians: float
		An angle in radians.

	Returns
	-------
	float
		The hyperbolic arccotangent of angle_in_radians in radians.

	Raises
	------
	ValueError
		If angle_in_radians is between -1 and 1.

	ZeroDivisionError
		If angle_in_radians is 1.
	"""
	return 0.5 * math.log((angle_in_radians + 1.0) / (angle_in_radians - 1.0))

File: java/src/TrigReciprocal.java
/**
 * Trigonometric reciprocal identity methods that return radians.
 * 
 * @author Synthird
 * @version 1.4.0
 */
public final class TrigReciprocal {

	/**
	 * TrigReciprocal class cannot be instantiated.
	 * 
	 * @throws IllegalAccessException
	 */
	public TrigReciprocal() throws IllegalAccessException {
		throw new IllegalAccessException("TrigReciprocal cannot be instantiated!");
	}

	/**
	 * The reciprocal of sin.
	 * 
	 * @param angleInRadians An angle in radians.
	 * @return The cosecant of angleInRadians in radians.
	 */
	public static final double csc(double angleInRadians) {
		return angleInRadians == 0.0 ? Double.NaN : 1.0 / Math.sin(angleInRadians);
	}

	/**
	 * The reciprocal of cos.
	 * 
	 * @param angleInRadians An angle in radians.
	 * @return The secant of angleInRadians in radians.
	 */
	public static final double sec(double angleInRadians) {
		return 1.0 / Math.cos(angleInRadians);
	}

	/**
	 * The reciprocal of tan.
	 * 
	 * @param angleInRadians An angle in radians.
	 * @return The cotangent of angleInRadians in radians.
	 */
	public static final double cot(double angleInRadians) {
		return angleInRadians == 0.0 ? Double.NaN : 1.0 / Math.tan(angleInRadians);
	}

	/**
	 * The inverse of csc.
	 * 
	 * @param angleInRadians An angle in radians.
	 * @return The arccosecant of angleInRadians in radians.
	 */
	public static final double acsc(double angleInRadians) {
		return Math.asin(1.0 / angleInRadians);
	}

	/**
	 * The inverse of sec.
	 * 
	 * @param angleInRadians An angle in radians.
	 * @return The arcsecant of angleInRadians.
	 */
	public static final double asec(double angleInRadians) {
		return Math.acos(1.0 / angleInRadians);
	}

	/**
	 * The inverse of cot.
	 * 
	 * @param angleInRadians An angle in radians.
	 * @return The arccotangent of angleInRadians.
	 */
	public static final double acot(double angleInRadians) {
		return Math.atan(1.0 / angleInRadians);
	}

	/**
	 * The hyperbolic version of csc.
	 * 
	 * @param angleInRadians An angle in radians.
	 * @return The hyperbolic cosecant of angleInRadians in radians.
	 */
	public static final double csch(double angleInRadians) {
		return angleInRadians == 0.0 ? Double.NaN : 1.0 / Math.sinh(angleInRadians);
	}

	/**
	 * The hyperbolic version of sec.
	 * 
	 * @param angleInRadians An angle in radians.
	 * @return The hyperbolic secant of angleInRadians in radians.
	 */
	public static final double sech(double angleInRadians) {
		return 1.0 / Math.cosh(angleInRadians);
	}

	/**
	 * The hyperbolic version of cot.
	 * 
	 * @param angleInRadians An angle in radians.
	 * @return The hyperbolic cotangent of angleInRadians in radians.
	 */
	public static final double coth(double angleInRadians) {
		return angleInRadians == 0.0 ? Double.NaN : 1.0 / Math.tanh(angleInRadians);
	}

	/**
	 * The inverse of csch.
	 * 
	 * @param angleInRadians An angle in radians.
	 * @return The hyperbolic arccosecant of angleInRadians in radians.
	 */
	public static final double acsch(double angleInRadians) {
		return Math.log((1.0 / angleInRadians) + (Math.sqrt(1.0 + Math.pow(angleInRadians, 2)) / Math.abs(angleInRadians)));
		// return angleInRadians == 0.0 ? Double.NaN
		// 		: Math.log(Math.sqrt((1.0 / Math.pow(angleInRadians, 2)) + 1.0) + (1.0 / angleInRadians));
	}

	/**
	 * The inverse of sech.
	 * 
	 * @param angleInRadians An angle in radians.
	 * @return The hyperbolic arcsecant of angleInRadians in radians.
	 */
	public static double asech(double angleInRadians) {
		return Math.log((1.0 + Math.sqrt(1.0 - Math.pow(angleInRadians, 2))) / angleInRadians);
	}

	/**
	 * The inverse of coth.
	 * 
	 * @param angleInRadians An angle in radians.
	 * @return The hyperbolic arccotangent of angleInRadians in radians.
	 */
	public static double acoth(double angleInRadians) {
		return 0.5 * Math.log((angleInRadians + 1.0) / (angleInRadians - 1.0));
	}
}
